### PR TITLE
Improved support for internationalised domains

### PR DESCRIPTION
**Pull request by snh**
I came across a couple of issues with [internationalised domain names](https://en.wikipedia.org/wiki/Internationalized_domain_name) today that I hope to solve in this PR.

---

PublicSuffix doesn't support Punycode encoded internationalised public suffixes, and instead expects them to be Unicode encoded (weppos/publicsuffix-ruby#46). If you feed in a Punycode encoded   public suffix, then no result, or a more broader result, may be returned.

Impacts:

- `Domain.valid_domain?` can return `false`, leading to almost everything in `Domain` failing, including `Domain.https_eligible?` always returning `false`.
- `CAA.records` isn't able to retrieve the CAA records for the apex domain.

---

Unfortunately Dnsruby is the reverse way around, and expects the internationalised domain to be Punycode encoded, and exceptions if fed a Unicode encoded internationalised domain (alexdalitz/dnsruby#94).

Impacts:

- Feeding a Unicode encoded internationalised domain in to most of the `GitHubPages::HealthCheck::Domain` or `GitHubPages::HealthCheck::CAA` methods will lead to a Dnsruby exception.

--- 

To work around these issues, I am feeding the domain in to Addressable's `normalized_host` method, which converts the internationalised domain to Punycode encoding, which appears to be the generally more accepted way to work with domains in code. I am then re-encoding in Unicode where required (which is just when using PublicSufix for now).

Addressable's `normalized_host` method will also take care of removing a trailing dot, making our existing way to doing this redundant.

These encoding transformations cope fine with a non-internationalised domain.

I have added tests, though these actually pass even without these fixes, as to surface this bug you need to use a domain with internationalised characters in the public suffix. I didn't want to pick a random name, since it may be in use, or registered in the future. It might be worth us registering a domain in one of the [internationalised country code top-level domain's](https://en.wikipedia.org/wiki/Internationalized_country_code_top-level_domain) which appears in the public suffix list.

__Glossary__:

__Public Suffix__: As defined at https://publicsuffix.org/, an example is the `com` portion of `pages.github.com`.
__Apex Domain__: As defined at https://help.github.com/articles/about-supported-custom-domains/#apex-domains, an example is the `github.com portion of `pages.github.com`.

### CODE REVIEW

File: spec/github_pages_health_check/domain_spec.rb [hunk 1]
```diff
@@ -859,11 +897,52 @@
...
+
+          it { is_expected.to be_https_eligible }
+        end
+      end
     end
 
     context "CNAME record pointed to username" do
       let(:cname) { "foobar.github.io" }
       before(:each) { allow(subject).to receive(:dns) { [cname_packet] } }
+      before(:each) { allow(subject.send(:caa)).to receive(:query) { [cname_packet] } }
```
Comment (snh) [inline spec/github_pages_health_check/domain_spec.rb]: This is required so that we return the `CNAME` when no `CAA` exists, imitating the behaviour seen in the real world. This hasn't been a issue until now, as the existing checks use a `domain` of `pages.github.com`, so we get back entries from the real world here. It is only now that I have started using fake names that this has become a problem.

File: spec/github_pages_health_check/domain_spec.rb [hunk 2]
```diff
@@ -836,13 +860,27 @@
...
+        let(:domain) { "xn--dmain-0ta.example.com" }
+
+        it { is_expected.not_to be_https_eligible }
+      end
     end
 
     context "A records pointed to new IPs" do
       let(:ip) { "185.199.108.153" }
       before(:each) { allow(subject).to receive(:dns) { [a_packet] } }
+      before(:each) { allow(subject.send(:caa)).to receive(:query) { [a_packet] } }
```
Comment (snh) [inline spec/github_pages_health_check/domain_spec.rb]: This is required so that we return the `A` when no `CAA` exists, imitating the behaviour seen in the real world. This hasn't been a issue until now, as the existing checks use a `domain` of `pages.github.com`, so we get back entries from the real world here. It is only now that I have started using fake names that this has become a problem.

File: spec/github_pages_health_check/domain_spec.rb [hunk 3]
```diff
@@ -880,11 +959,52 @@
...
+
+          it { is_expected.to be_https_eligible }
+        end
+      end
     end
 
     context "CNAME record pointed elsewhere" do
       let(:cname) { "jinglebells.com" }
       before(:each) { allow(subject).to receive(:dns) { [cname_packet] } }
+      before(:each) { allow(subject.send(:caa)).to receive(:query) { [cname_packet] } }
```
Comment (snh) [inline spec/github_pages_health_check/domain_spec.rb]: This is required so that we return the `CNAME` when no `CAA` exists, imitating the behaviour seen in the real world. This hasn't been a issue until now, as the existing checks use a `domain` of `pages.github.com`, so we get back entries from the real world here. It is only now that I have started using fake names that this has become a problem.

File: spec/github_pages_health_check/domain_spec.rb [hunk 4]
```diff
@@ -836,13 +860,27 @@
     context "A records pointed to old IPs" do
       let(:ip) { "192.30.252.153" }
       before(:each) { allow(subject).to receive(:dns) { [a_packet] } }
+      before(:each) { allow(subject.send(:caa)).to receive(:query) { [a_packet] } }
```
Comment (snh) [inline spec/github_pages_health_check/domain_spec.rb]: This is required so that we return the `A` when no `CAA` exists, imitating the behaviour seen in the real world. This hasn't been a issue until now, as the existing checks use a `domain` of `pages.github.com`, so we get back entries from the real world here. It is only now that I have started using fake names that this has become a problem.

File: lib/github-pages-health-check/domain.rb [hunk 5]
```diff
@@ -100,7 +100,7 @@ def initialize(host, nameservers: :default)
 
         @host = normalize_host(host)
         @nameservers = nameservers
-        @resolver = GitHubPages::HealthCheck::Resolver.new(host,
+        @resolver = GitHubPages::HealthCheck::Resolver.new(@host,
```
Comment (snh) [inline lib/github-pages-health-check/domain.rb]: Found this while undertaking some more testing just then when I managed to trigger some further Dnsruby exceptions when using a Unicode encoded IDN. Seems we weren't passing the normalised host through to the `Resolver` class.

Comment (parkr) [reply lib/github-pages-health-check/domain.rb]: The `host` method should have been normalized?

Comment (snh) [reply lib/github-pages-health-check/domain.rb]: As `host` exists as a locally scoped variable in the `initialize` method, my understanding is that we need to explicitly refer to the instance level scoped `host` where we want to use it, otherwise we will get back the locally scoped one, which is pre `normalize_host`.

```
$ cat local-vs-instance.rb
def test(host)
  @host = host.upcase
  puts "host: #{host}, @host: #{@host}"
end

test("github.com")

$ ruby local-vs-instance.rb
host: github.com, @host: GITHUB.COM
```

Comment (parkr) [reply lib/github-pages-health-check/domain.rb]: Totally right! The `host` parameter is first in the order of precedence, then the interpreter will attempt to find a method called `host`. I'd love to use `self.host` here so that we're calling the method. I'm worried about reordering these calls and causing `@host` to be `nil` by accident. What do you think?

Comment (snh) [reply lib/github-pages-health-check/domain.rb]: Thanks @parkr, I'm a little lost here, as from what I understand we don't have a `host` method?, would you mind expanding on what you mean here?

Comment (parkr) [reply lib/github-pages-health-check/domain.rb]: On line 6, there's an `attr_reader` for `:host` which creates a reader for `@host`! That's where I'm understanding that there is a host method. 😄

Comment (snh) [reply lib/github-pages-health-check/domain.rb]: Thanks @parkr, it finally clicked 😄 I've updated the PR to use `self.host`.
